standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java

```diff
@@ -20,71 +20,666 @@
+ * This class is a singleton, which is instantiated the first time
+ * method determineDatabaseProduct is invoked.
+ * Tests that need to create multiple instances can use the reset method
+ * */
+public class DatabaseProduct implements Configurable {
+  static final private Logger LOG = LoggerFactory.getLogger(DatabaseProduct.class.getName());
+
+  private static enum DbType {DERBY, MYSQL, POSTGRES, ORACLE, SQLSERVER, CUSTOM, UNDEFINED};
+  public DbType dbType;
+  
+  // Singleton instance
+  private static DatabaseProduct theDatabaseProduct;
+
+  Configuration myConf;
+  /**
+   * Protected constructor for singleton class
+   * @param id
+   */
+  protected DatabaseProduct() {}
+
+  public static final String DERBY_NAME = "derby";
+  public static final String SQL_SERVER_NAME = "microsoft sql server";
+  public static final String MYSQL_NAME = "mysql";
+  public static final String POSTGRESQL_NAME = "postgresql";
+  public static final String ORACLE_NAME = "oracle";
+  public static final String UNDEFINED_NAME = "other";
+  
   /**
    * Determine the database product type
    * @param productName string to defer database connection
    * @return database product type
    */
-  public static DatabaseProduct determineDatabaseProduct(String productName) throws SQLException {
-    if (productName == null) {
-      return OTHER;
+  public static DatabaseProduct determineDatabaseProduct(String productName, Configuration c) {
+    DbType dbt;
+
+    if (theDatabaseProduct != null) {
+      Preconditions.checkState(theDatabaseProduct.dbType == getDbType(productName));
+      return theDatabaseProduct;
     }
+
+    // This method may be invoked by concurrent connections
+    synchronized (DatabaseProduct.class) {
+
+      if (productName == null) {
+        productName = UNDEFINED_NAME;
+      }
+
+      dbt = getDbType(productName);
+
+      // Check for null again in case of race condition
+      if (theDatabaseProduct == null) {
+        final Configuration conf = c!= null ? c : MetastoreConf.newMetastoreConf();
+        // Check if we are using an external database product
+        boolean isExternal = MetastoreConf.getBoolVar(conf, ConfVars.USE_CUSTOM_RDBMS);
+
+        if (isExternal) {
+          // The DatabaseProduct will be created by instantiating an external class via
+          // reflection. The external class can override any method in the current class
+          String className = MetastoreConf.getVar(conf, ConfVars.CUSTOM_RDBMS_CLASSNAME);
+          
+          if (className != null) {
+            try {
+              theDatabaseProduct = (DatabaseProduct)
+                  ReflectionUtils.newInstance(Class.forName(className), conf);
+  
+              LOG.info(String.format("Using custom RDBMS %s. Overriding DbType: %s", className, dbt));
+              dbt = DbType.CUSTOM;
+            }catch (Exception e) {
+              LOG.warn("Caught exception instantiating custom database product. Reverting to " + dbt, e);
```

Comment:
Is there a good reason to catch the exception and reverting the dbt? Should'nt we throw a MetaException so that HMS doesn't start if the configuration is invalid.

Reply:
I changed the code to throw a RuntimeException instead. This method is called in a few places where it's not clear what to do with a regular Exception. Hope this makes sense.